docs/logging-design.md

```diff
@@ -0,0 +1,51 @@
+# Design of the way AsyncHTTPClient logs
+
+<details>
+  <summary>Unless you explicitly pass AsyncHTTPClient a `Logger` instance, nothing is ever logged.</summary>
+  AsyncHTTPClient is useful in many places where you wouldn't want to log, for example a command line HTTP client. Also, we do not want to change its default behaviour in a minor release.
+</details>
+<details>
+  <summary>Nothing is logged at level `info` or higher, unless something is really wrong that cannot be communicated through the API.</summary>
+  Fundamentally, AsyncHTTPClient performs a simple task, it makes a HTTP request and communicates the outcome back via its API. In normal usage, we would not expect people to want AsyncHTTPClient to log. In certain scenarios, for example when debugging why a request takes longer than expected it may however be useful to get information about AsyncHTTPClient's connection pool. That is when enabling logging may become useful.
+</details>
+<details>
+  <summary>Each request will get a globally unique request ID (`ahc-request-id`) that will be attached (as metadata) to each log message relevant to a request.</summary>
```

Comment:
Shouldn't we offload the burden of setting unique `ahc-request-id`'s to a client? We already accept a logger for every method, for the client it'll be extremely simple to set a unique uuid in it. UUID creation is not free (and when passed in metadata it's actually materialised unless some custom code is written), I think we'll be fine with just giving client a way to set a unique uuid per http request.

Reply:
Such "at beginning inject some metadata to context (logger metadata)" is what the baggage context infrastructure is being designed for. 

Specifically "if no trace header detected, make a new span id and store it" is the most common implementation of a tracer performing an "extract (from headers to context). So that's something a tracer / baggage context "instrument" (we don't have settled names for those yet) would be doing...

If you want to do something ad hoc for now that's also ok I guess, but let's keep an eye out for the overlap here and how the context infra would be used to achieve the same (and more, trace header extraction) in the future.

Reply:
this is a really good point - we should design this feature with an eye towards the distributed tracing work we kicked off in GSoC. to be clear, we should be careful to hold back progress here, but we should design this in a way  that can be evolved into full distributed tracing

Reply:
I think we might be okey with doing "whatever is easiest" here for the time being and once the baggage baggage context matures I'm expecting it'd be medium effort to plug the right things anyway, and likely not "just" change one generator for another. 

So let's just keep in mind that this part is very likely to change, and let's do so when the other lib is ready.

// cc @slashmo

Reply:
I may be missing something, but why do we ever want to generate ids inside the library?
It can be achieved with a wrapper (since it can take `Logger`, inject some variables and pass it to a delegate). So does the tracing or whatever keys we want to pass inside. Basically, we can provide a wrapper that solves the problem and shelve this work until we know how tracing API will look like.

Reply:
@avolokhov / @ktoso / @tomerd The reason I went ahead adding a `ahc-request-id` (which is just an atomically incrementing `Int`) is that without it, correlating the log messages is impossible. Sure, the user _can_ add an ID but if without adding one in AHC, they _have to_ do it to make sense of the log messages. I thought that's too much and therefore I thought it may be useful to just have an ID for AHC.